### ISSUE TITLE
ModelValidationError when building TokenGroupBase w/out --simple

### ISSUE THREAD

**Issue by seizethedave** [CozySynthesizer/cozy#14]
(As of v2 commit 86d09a2c77b0ac7706c7a6d2a0ab7716ff0e2f65)

Spec:
https://github.com/uwplse/cozy/blob/dogfood/specs/TokenGroupBase.ds

Command:
`cozy --java out.java specs/TokenGroupBase.ds`

Error:
```
Traceback (most recent call last):
  File "/Users/dgrant/Dev/cozy/cozy/cozy/jobs.py", line 30, in _run
    self.run()
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/high_level_interface.py", line 102, in run
    stop_callback=lambda: self.stop_requested)):
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/core.py", line 642, in improve
    if eliminate_vars.value and can_elim_vars(target, assumptions, state_vars):
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/core.py", line 555, in can_elim_vars
    EEq(spec, subst(spec, sub))))
  File "/Users/dgrant/Dev/cozy/cozy/cozy/solver.py", line 866, in valid
    return satisfy(EUnaryOp("not", e).with_type(BOOL), **opts) is None
  File "/Users/dgrant/Dev/cozy/cozy/cozy/solver.py", line 859, in satisfy
    raise ModelValidationError("model validation failed")
cozy.solver.ModelValidationError: model validation failed
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/bin/cozy", line 11, in <module>
    load_entry_point('Cozy', 'console_scripts', 'cozy')()
  File "/Users/dgrant/Dev/cozy/cozy/cozy/main.py", line 104, in run
    progress_callback = callback)
  File "/Users/dgrant/Dev/cozy/cozy/cozy/common.py", line 59, in g
    ret = f(*args, **kwargs)
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/high_level_interface.py", line 183, in improve_implementation
    raise Exception("failed job: {}".format(j))
Exception: failed job: ImproveQueryJob[isDistinct]
```

**Comment by Calvin-L** [comment]
We will probably have to work through a few of these.

You can determine that the query `isDistinct` caused the failure from this line:
```Exception: failed job: ImproveQueryJob[isDistinct]```

When this happens, a lot of debug info gets written to `/tmp/isDistinct.log`. Could you include that file as an attachment?

**Comment by seizethedave** [comment]
Oh yes, here:

[isDistinct.log](https://github.com/uwplse/cozy/files/1438394/isDistinct.log)

**Comment by Calvin-L** [comment]
Fixed this particular issue in 2f35754db4c536c31001ceeffd231ccf56b06783. There might be others.
